chainer/functions/normalization/batch_normalization.py

```diff
@@ -152,205 +395,69 @@ def forward(self, inputs):
-                    self.running_var.inplace_axpby(
-                        self.decay, (1 - self.decay), self.var * adjust)
-                else:
-                    self.running_var *= self.decay
-                    self.running_var += self.var * adjust * (1 - self.decay)
-
-        elif self.use_cudnn:
-            if self.running_mean is not None:
-                mean = self.running_mean
-                var = self.running_var
-            else:
-                # Create dummies.
-                mean = xp.zeros_like(gamma, dtype=x.dtype)
-                var = xp.zeros_like(gamma, dtype=x.dtype)
-
-            # self.mean and self.inv_std are used as buffers to save
-            # intermediate results computed during forward pass. These buffers
-            # are used to speed-up backward pass.
-            y, self.mean, self.inv_std = (
-                cudnn.batch_normalization_forward_training(
-                    x, gamma, beta, mean, var, None, None,
-                    self.eps, self.decay, self.mode.is_for_conv2d,
-                    self.mode.get_cudnn_mode(), chainer.is_debug()))
-        else:
-            # Generic CPU and GPU implementation
-
-            interm_dtype = numpy.promote_types(x.dtype, gamma.dtype)
-
-            gamma = gamma[expander].astype(interm_dtype, copy=False)
-            beta = beta[expander].astype(interm_dtype, copy=False)
-            self.mean = x.mean(axis=self.axis, dtype=interm_dtype)
-            var = x.var(axis=self.axis, dtype=interm_dtype)
-            if xp is numpy:
-                self.inv_std = numpy.reciprocal(numpy.sqrt(
-                    var + self.eps, dtype=interm_dtype))
-            else:
-                self.inv_std = cuda.cupyx.rsqrt(
-                    var + self.eps, dtype=interm_dtype)
 
-            y = _apply_bn_fwd(xp, x, self.mean[expander],
-                              self.inv_std[expander], gamma, beta)
-
-            # Update running statistics if given
-            if self.running_mean is not None:
-                m = x.size // gamma.size
-                adjust = m / max(m - 1., 1.)  # unbiased estimation
-
-                xp = backend.get_array_module(
-                    self.running_mean, self.running_var)
-
-                if xp is chainerx:
-                    self.running_mean, self.running_var = backend.from_chx(
-                        (self.running_mean, self.running_var))
-
-                self.running_mean *= self.decay
-                self.running_mean += (1 - self.decay) * self.mean
-                self.running_var *= self.decay
-                self.running_var += (1 - self.decay) * adjust * var
-
-                if xp is chainerx:
-                    self.running_mean = backend.to_chx(self.running_mean)
-                    self.running_var = backend.to_chx(self.running_var)
+        self.mode = _BNMode(x, gamma, self.key_axis)
+        self.bn_backend = self.bn_backend_selector(xp, self.mode)
```

Comment:
Rather than passing `self.mode` to the selector, how about passing `self` and `inputs`?
`_BNMode` is a highly internal entity, not appropriate for interfacing.

Reply:
I am not sure I understand. Currently, we need `can_use_cudnn`, `can_use_ideep`, `is_for_conv2d`, `get_cudnn_mode` from `mode`. Do you mean we put the code to calculate these four into the selector?
https://github.com/chainer/chainer/blob/5080ad1c4faa0caf8cc14fea1a07dde30e066fe7/chainer/functions/normalization/batch_normalization.py#L265

Reply:
As the `mode` is removed from member in 8be6e49. Passing `self` cannot have access the `mode`

Reply:
`_BNMode` can be created inside `_bn_backend_selector` from `self` and `inputs`, as it's not used in `multinode_bn_backend_selector`.

Reply:
I see, I thought you mean we should not call anything related to `_BNMode` from outside of `BatchNormalization`.

Reply:
I made a draft in d218209. Actually, we only need the `key_axis` from `BatchNormalization` to create `_BNMode`. But I thought what would be too limited, what do you think.

Reply:
Currently only `key_axis` is required because of cuDNN's specification. I don't think it's a good idea to encode that into the interface.
We could pass all the possibly-useful attributes, but I think passing `BatchNormalization` is an easier option for now.

Reply:
Thanks. Then I think d218209 has addressed the issue